Paternity Test.py

```diff
@@ -0,0 +1,378 @@
+        try:
+            if(df.at[index_label,'FileType'] == PARENT_FILETYPE):
+
+                # Find this parent's child
+                # Child might be above parent so we can't just start from index of the parent
+                for index_label_inception, row_series_inception in df.iterrows():
+                    logging.debug("Outer iter %d, inner iter %d parent %s checking if %s is its child" % (index_label, index_label_inception, df.at[index_label,'filename'], df.at[index_label_inception,'filename']))
+                    if index_label != index_label_inception: #needed to preventing it find itself
+                        if(df.at[index_label,'ID'] == #if parent ID...
+                           df.at[index_label_inception,'ID']) and ( #equals potential child ID, and...
+                            df.at[index_label_inception,'FileType'] == #child's file extension...
+                            child_extension): #equals the current child_extension we are iterating on
+                            # Child found!
+                            logging.debug("    Found "+df.at[index_label_inception,'filename']+" to be child")
+                            df.at[index_label,child_extension] = df.at[index_label_inception,'filename']
+                            df.at[index_label,child_extension+"_location"] = df.at[index_label_inception,'location']
+                            df.drop([index_label_inception], inplace=True)
+                            index_label = index_label+1 # Adjust index accordingly
+                            break
+
+                # We were unable to find a child for this parent
+                if(df.at[index_label,child_extension] == ""):
+                    logging.warning("Could not find child of type %s for parent %s" % (child_extension, df.at[index_label, 'FileType']))
+                    df.at[index_label,child_extension] = ""
+                    df.at[index_label,child_extension+"_location"] = ""
+        except KeyError:
+            pass
+logging.info("Cleaning up dataframe...")
+# Iterate one more time to delete child rows
+# Because children could appear above their parents, deleting during the above iteration could mess things up
+df.drop(columns=['ID'], inplace=True)
+df.rename(columns = {'FileType':'parent_file_ext'}, inplace = True)
+for index_label, row_series in df.iterrows():
+        if(df.at[index_label,'parent_file_ext'] != PARENT_FILETYPE):
+            df.drop([index_label], inplace=True)
+df.reset_index(inplace=True, drop=True)
+logging.info("Finished")
+
+
+# ### Inspect dataframe
+
+# This is another optional step. Panadas will not display all of your rows if there are many of them.
+
+# In[ ]:
+
+
+print(df)
+
+
+# ### Generate TSV file from dataframe and upload it
+
+# In[ ]:
+
+
+#!rm final.tsv
+# ^ Uncomment above line if you will be running this block more than once
+
+df.to_csv("dataframe.tsv", sep='\t')
```

Comment:
You don't need to save this to a tsv, then open the tsv, then write it again to a tsv just to include your header.  See if you can write the dataframe + header directly to a file.

Reply:
At the time, this was the only way I could get around some issues Terra has with header strings, but I'll give it another go and see if I can make it cleaner.